src/EventStore.Core/TransactionLog/Chunks/TFChunk/WriterWorkItem.cs

```diff
@@ -1,90 +1,74 @@
-using System;
 using System.IO;
 using System.Security.Cryptography;
-using EventStore.Common.Utils;
+using DotNext;
+using DotNext.IO;
+using Microsoft.Win32.SafeHandles;
 
 namespace EventStore.Core.TransactionLog.Chunks.TFChunk {
-	internal class WriterWorkItem {
-		public Stream WorkingStream {
-			get { return _workingStream; }
-		}
-
-		public long StreamLength {
-			get { return _workingStream.Length; }
-		}
+	internal sealed class WriterWorkItem : Disposable {
+		public const int BufferSize = 8192;
 
-		public long StreamPosition {
-			get { return _workingStream.Position; }
-		}
+		public Stream WorkingStream { get; private set; }
 
 		private readonly Stream _fileStream;
 		private UnmanagedMemoryStream _memStream;
-		private Stream _workingStream;
 
 		public readonly MemoryStream Buffer;
 		public readonly BinaryWriter BufferWriter;
 		public readonly HashAlgorithm MD5;
 
-		public WriterWorkItem(Stream fileStream, UnmanagedMemoryStream memStream, HashAlgorithm md5) {
-			_fileStream = fileStream;
-			_memStream = memStream;
-			_workingStream = (Stream)fileStream ?? memStream;
-			Buffer = new MemoryStream(8192);
-			BufferWriter = new BinaryWriter(Buffer);
+		public unsafe WriterWorkItem(nint memoryPtr, int length, HashAlgorithm md5) {
+			WorkingStream = _memStream = new UnmanagedMemoryStream((byte*)memoryPtr, length, length, FileAccess.ReadWrite);
+			Buffer = new(BufferSize);
+			BufferWriter = new(Buffer);
+			MD5 = md5;
+		}
+
+		public WriterWorkItem(SafeFileHandle handle, HashAlgorithm md5, bool unbuffered) {
+			WorkingStream = _fileStream = unbuffered
+				? handle.AsUnbufferedStream(FileAccess.ReadWrite)
+				: new BufferedStream(handle.AsUnbufferedStream(FileAccess.ReadWrite), BufferSize);
+			Buffer = new(BufferSize);
+			BufferWriter = new(Buffer);
 			MD5 = md5;
 		}
 
 		public void SetMemStream(UnmanagedMemoryStream memStream) {
 			_memStream = memStream;
-			if (_fileStream == null)
-				_workingStream = memStream;
+			if (_fileStream is null)
+				WorkingStream = memStream;
 		}
 
 		public void AppendData(byte[] buf, int offset, int len) {
 			// as we are always append-only, stream's position should be right here
-			if (_fileStream != null)
-				_fileStream.Write(buf, 0, len);
+			_fileStream?.Write(buf, offset, len);
+
 			//MEMORY
-			var memStream = _memStream;
-			if (memStream != null)
-				memStream.Write(buf, 0, len);
+			_memStream?.Write(buf, offset, len);
 		}
 
 		public void ResizeStream(int fileSize) {
-			if (_fileStream != null)
-				_fileStream.SetLength(fileSize);
-			var memStream = _memStream;
-			if (memStream != null)
-				memStream.SetLength(fileSize);
+			_fileStream?.SetLength(fileSize);
+			_memStream?.SetLength(fileSize);
 		}
 
-		public void Dispose() {
-			if (_fileStream != null)
-				_fileStream.Dispose();
+		protected override void Dispose(bool disposing) {
+			if (disposing) {
+				_fileStream?.Dispose();
```

Comment:
is it guaranteed that dotNext's `UnbufferedFileStream` will flush the file to disk when being disposed? (the previous code implicitly assumes that it does)

Reply:
> is it guaranteed that dotNext's UnbufferedFileStream will flush the file to disk when being disposed

No, it's not. However, most of the code inside `TFChunk` do `Flush` before disposing of writer item. Redundant call to `FlushToDisk` leads to extra syscall with no effect. Maybe I missed something. If so, I can add `FlushToDisk` explicitly.

Reply:
the case i'm worried about is when doing a manual shut down of the database:
`bool TryClose()` in TFChunk should return true only if the chunk was properly closed and flushed if it was being written to (otherwise, the writer checkpoint may be flushed but not the chunk which would cause the database to become inconsistent - see this [bug](https://github.com/EventStore/EventStore/pull/3669) for example)

Based on online articles, it appears that .NET's Filestream's automatically flushes when it's being closed, so we previously didn't need to have an explicit Flush() call, but now it appears that we do need it.

Reply:
yeah good catch :+1: